### PR TITLE
Ariasanovsky/issue6

### PR DESCRIPTION
**Pull request by ariasanovsky**
I fleshed out the design needs of this crate through example/test-driven development, populating `$HOME/examples/` with experimental design patterns. I concluded with a data-oriented design with `rayon` operating on zipped arrays to perform operations in parallel.

In summary:

- `01-r33.rs`: toying with the $R(3, 3) > 5$ problem through a top-down view
  - useful to get rudimentary mechanics for tree simulations and updates
  - abandoned to accommodate neural network friendly batched evaluations
  - used traits to set up contracts (prematurely, as we see next)
- `02-r33.rs`: same math problem with an aggressive config typestate pattern and logging routines
  - the config was helpful for avoiding wide turbofish, but tedious to implement
  - the intent was to set up contracts for everything from reward function to granular performance gains (delaying `f32` and `i32` addition), and different upper estimate functions
  - it was too cumbersome to refactor due to the syntactic overhead
- `03-r44.rs`: toying with $R(4, 4) > 17$ with a focus on batched NN evaluations
  - starting with the NN guided us to the correct function signatures
  - however, this required another complete rewrite to the tree structure to facilitate the data-oriented needs
- `04-r44.rs`: same math problem but driven directly from a more concrete pattern
  - function signatures defined one-by-one so the learning loop can be fully written
  - this led to a focus on separate arrays for each struct: states, state vectors, trees, etc
  - state vector arrays are easy to pass into a NN, and observation vectors for computing loss
  - additionally, the `ramsey` subcrate was hard-coded with N = 17 to minimize turbofishing and frustration with the desired const generic features

In conclusion, the next issues will focus on taking the minimal steps needed to prove $R(4, 4) > 17$ in `04-r44.rs` with some logging. I would also love to incorporate my $(g, g, g)$ value vector soon.

### CODE REVIEW

**Comment by ariasanovsky** [comment]
I fleshed out the design needs of this crate through example/test-driven development, populating `$HOME/examples/` with experimental design patterns. I concluded with a data-oriented design with `rayon` operating on zipped arrays to perform operations in parallel.

In summary:

- `01-r33.rs`: toying with the $R(3, 3) > 5$ problem through a top-down view
  - useful to get rudimentary mechanics for tree simulations and updates
  - abandoned to accommodate neural network friendly batched evaluations
  - used traits to set up contracts (prematurely, as we see next)
- `02-r33.rs`: same math problem with an aggressive config typestate pattern and logging routines
  - the config was helpful for avoiding wide turbofish, but tedious to implement
  - the intent was to set up contracts for everything from reward function to granular performance gains (delaying `f32` and `i32` addition), and different upper estimate functions
  - it was too cumbersome to refactor due to the syntactic overhead
- `03-r44.rs`: toying with $R(4, 4) > 17$ with a focus on batched NN evaluations
  - starting with the NN guided us to the correct function signatures
  - however, this required another complete rewrite to the tree structure to facilitate the data-oriented needs
- `04-r44.rs`: same math problem but driven directly from a more concrete pattern
  - function signatures defined one-by-one so the learning loop can be fully written
  - this led to a focus on separate arrays for each struct: states, state vectors, trees, etc
  - state vector arrays are easy to pass into a NN, and observation vectors for computing loss
  - additionally, the `ramsey` subcrate was hard-coded with N = 17 to minimize turbofishing and frustration with the desired const generic features

In conclusion, the next issues will focus on taking the minimal steps needed to prove $R(4, 4) > 17$ in `04-r44.rs` with some logging. I would also love to incorporate my $(g, g, g)$ value vector soon.
